### ISSUE TITLE
False negative on UnsafeCast

### ISSUE THREAD

**Issue by dive5617** [arturbosch/detekt#2082]
```kotlin
import java.sql.DriverManager

open class Father {
    open val key: String = "hello"
}

class Child : Father() {
    override val key: String = "world"
}

fun main() {
    val mp: Map<String, Father>? = null
    val key = (mp?.get("1") as Child).key
}
```

The `as Child` could be nullable but `UnsafeCast` didn't report any warning.

It looks like this https://github.com/arturbosch/detekt/pull/1879 caused this bug @3flex 

## Expected Behavior

The rule will work and the detekt will report an error.

## Observed Behavior

Detekt does not report an error.

## Context

```
potential-bugs:
  active: true
  UnsafeCast:
    active: true
```

## Your Environment

* Version of detekt used: 1.1.1
* Version of Gradle used (if applicable): use cmd
* macOS Catalina Version 10.15

**Comment by BraisGabin** [comment]
I just reproduced it with this code:

```kotlin
val map: Map<String, String>? = null
val value = map?.get("1") as String
```

**Comment by arturbosch** [comment]
This rule does not work anymore without the experimental `--classpath` flag/property.
You can try running `gradle detektMain` or `gradle detektTest` in the corresponding module to have this rule enabled as the source set specified tasks use the classpath.
Note that the classpath/type resolution features are not yet documented.

**Comment by dive5617** [comment]
Thanks for your kindly reply.  @arturbosch 
I test single kotlin file with cmd include related `--classpath` and `Jvm` flags and ensure this rule is enable since other cases like 
```val something: String = 3 as String```
are detected successfully by `UnsafeCast`.

**Comment by dive5617** [comment]
So is it a bug or expected behavior?

**Comment by dive5617** [comment]
@arturbosch I wonder whether it's a bug because this rule is enabled but nothing report.

**Comment by 3flex** [comment]
As mentioned on #1879 there will be some false negatives now the Kotlin compiler rules are being used instead of a custom detection in detekt.

Unfortunately the custom rules had too many false positives which is why that change was made, but some false negatives are now expected. We can always try and improve this in future.

So assuming the `--classpath` flag is set correctly this isn't a support issue but actually a false negative, which we don't have a label for... Should we add one @arturbosch?

**Comment by dive5617** [comment]
I think `UnsafeCast` has been  enable successfully because other unsafe cases I mentioned above were reported.

**Comment by xouabita** [comment]
It seems that `UnsafeCast` report the same as the kotlin compiler warning `CAST_NEVER_SUCCEEDS`:

code:
```kotlin
fun main() {
    val foo = "bar"
    println(foo as Int)
}
```

command:
```bash
kotlinc test.kt -include-runtime
```

output:
```bash
test.kt:3:17: warning: this cast can never succeed
    println(foo as Int)
```

Which make this rule redundant if you already have warnings as errors.

Would it be possible to bring back the old behavior of this rule ? When it was triggering for every `as` ? Maybe behind a preference in `config.yml` ?

**Comment by arturbosch** [comment]
@3flex sure, false negative sounds good to me to have.

@xouabita such a config property sounds also good to me. I've heard some users do not want any `as` in their code base.

**Comment by schalkms** [comment]
Should we even do anything here? We comply with IntelliJ's behavior. I wouldn't want to include any extra logic which diverges from the original behavior.

**Comment by arturbosch** [comment]
@schalkms as we just delegate to the compiler, we can't do anything about it but report it to the Kotlin team.
